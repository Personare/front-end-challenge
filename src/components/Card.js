import React, { PropTypes } from 'react';
import TweenLite from 'gsap';

export class Card extends React.Component {
    componentWillReceiveProps(nextProps) {
        if (nextProps.selected !== this.props.selected) {
            TweenLite.to(this.refs.cardFaces, 1, { rotationY: !nextProps.selected ? -180 : 0 });
        }
    }

    render() {
        return (
            <article className="card" onClick={this.props.onClick}>
                <div className="card-faces" ref="cardFaces">
                    <div className="front">
                        <img src={this.props.frontImg} alt={this.props.name} />
                    </div>
                    <div className="back">
                        <img src={this.props.backImg} alt={this.props.name} />
                    </div>
                </div>
            </article>
        );
    }
}

Card.propTypes = {
    name: PropTypes.string,
    frontImg: PropTypes.string,
    backImg: PropTypes.string,
    selected: PropTypes.bool,
    onClick: PropTypes.func,
};
