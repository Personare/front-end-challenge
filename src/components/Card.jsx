import React from "react";

const Card = ({ name = "CartjÃªnha", image = "", ...rest }) => {
  return (
    <div className="card">
      <img src={image} alt={name} />
    </div>
  );
};

export default Card;
