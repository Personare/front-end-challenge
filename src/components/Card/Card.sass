.CardContainer
  position: relative
  height: 210px
  cursor: pointer
  margin: 10px;
  .CardContent
    transform-style: preserve-3d

  .CardImage, .BackImage
    position: absolute
    width: 100px
    height: 210px
    overflow: hidden
    border-radius: 5px
    backface-visibility: hidden
    > img
      width: 100px
      height: 210px

  .BackImage
    transform: rotateY(-180deg)
