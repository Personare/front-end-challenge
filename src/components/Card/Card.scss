.Card {
    position: relative;
    perspective: 150rem;
    width: 200px;
    height: 200px;

    &__side {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 50px;
        backface-visibility: hidden;
        transition: all .8s ease;

        &.--front,
        &.--back {
            svg {
                position: absolute;
                top: 0;
                right: 0;
                bottom: 0;
                left: 0;
                width: 1em;
                height: 1em;
                font-size: 5.4rem;
            }
        }

        &.--front {
            background-color: red;

            svg {
                fill: black;
            }
        }

        &.--back {
            background-color: blue;
            transform: rotateY(180deg);

            svg {
                fill: red;
            }
        }
    }

    &.--closed &__side.\--front {
        transform: rotateY(-180deg);
    }

    &.--closed &__side.\--back {
        transform: rotateY(0);
    }
}

