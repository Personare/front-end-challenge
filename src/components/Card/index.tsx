import { useContext } from "react";
import ReactCardFlip from "react-card-flip";
import { ICard } from "../../config/interfaces";
import GlobalContext from "../../global/GlobalContext";

interface ICardItem {
  card: ICard;
  onClick?: (name: string) => void;
}

const Card: React.FC<ICardItem> = ({ card, onClick }) => {
  const { name, image, show } = card;

  const {
    state: { data },
  }: any = useContext(GlobalContext);

  const isFlipped = !show;

  console.log(onClick);

  return (
    <div data-cy="card" onClick={() => onClick && onClick(name)}>
      <ReactCardFlip isFlipped={isFlipped} flipDirection="horizontal">
        <img
          alt={name}
          src={`${data.imagesUrl}${image}`}
          data-cy={!isFlipped && "card-up"}
        />

        <img
          alt={name}
          src={data.imageBackCard}
          data-cy={isFlipped && "card-down"}
        />
      </ReactCardFlip>
    </div>
  );
};

export default Card;
