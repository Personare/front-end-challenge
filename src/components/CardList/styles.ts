import styled from "styled-components";

export const List = styled.div`
  align-items: center;
  display: grid;
  gap: 20px;
  width: 100%;

  flex-direction: column;
  grid-template-columns: repeat(auto-fill, minmax(162px, 1fr));

  img {
    display: block;
    height: auto;
    width: 100%;
  }
`;
