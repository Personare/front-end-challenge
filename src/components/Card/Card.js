import React, { Component } from 'react';
import TweenMax from 'gsap';

import './Card.sass';

export class Card extends Component {
  constructor(props) {
    super(props);
    this.cardContentRef = React.createRef();
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.selected !== this.props.selected) {
      TweenMax.to(this.cardContentRef.current, 1, { rotationY: !nextProps.selected ? -180 : 0 });
    }
  }

  render() {
    return (
      <article className="CardContainer" onClick={this.props.click}>
        <div className="CardContent" ref={this.cardContentRef}>
          <div className="CardImage">
            <img src={this.props.card} alt={this.props.name} />
          </div>
          <div className="BackImage">
            <img src={this.props.imageBack} alt={this.props.name} />
          </div>
        </div>
      </article>
    );
  }
}

export default Card;
