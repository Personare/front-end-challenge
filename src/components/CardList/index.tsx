import { ICard } from "../../config/interfaces";
import Card from "../Card";
import { List } from "./styles";

interface ICardList {
  cards: ICard[];
  onClick?: (name: string) => void;
}

const CardList: React.FC<ICardList> = ({ cards, onClick }) => {
  return (
    <List data-cy="card-list">
      {cards?.map((card: ICard) => (
        <Card key={card.name} card={card} onClick={onClick} />
      ))}
    </List>
  );
};
export default CardList;
