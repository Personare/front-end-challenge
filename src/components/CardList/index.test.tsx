import { shallow } from "enzyme";
import CardList from ".";
import Card from "../Card";

const CARDS = [
  {
    name: "Name Test #1",
    image: "Image Test #1",
    show: false,
  },
  {
    name: "Name Test #2",
    image: "Image Test #2",
    show: false,
  },
  {
    name: "Name Test #3",
    image: "Image Test #3",
    show: false,
  },
];

describe("CardList", () => {
  it("should match snapshot", () => {
    const wrapper = shallow(<CardList cards={CARDS} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render three items", () => {
    const wrapper = shallow(<CardList cards={CARDS} />);
    expect(wrapper.find(Card)).toHaveLength(CARDS.length);
  });
});
