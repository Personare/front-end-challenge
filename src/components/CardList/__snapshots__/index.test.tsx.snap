// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList should match snapshot 1`] = `
<styled.div>
  <Card
    card={
      Object {
        "image": "Image Test #1",
        "name": "Name Test #1",
        "show": false,
      }
    }
    key="Name Test #1"
  />
  <Card
    card={
      Object {
        "image": "Image Test #2",
        "name": "Name Test #2",
        "show": false,
      }
    }
    key="Name Test #2"
  />
  <Card
    card={
      Object {
        "image": "Image Test #3",
        "name": "Name Test #3",
        "show": false,
      }
    }
    key="Name Test #3"
  />
</styled.div>
`;
